### PR TITLE
Improve Textarea Behavior

### PR DESCRIPTION
**Pull request by Sukhendu2002**
**Changes**
This PR addresses issues related to the custom textarea behavior:
- Prevents the cursor from jumping to the top of the screen when interacting with the text area.
- Resizes the text area when content is removed, ensuring a better user experience.

**Issues**
Fixes https://github.com/jellyfin/jellyfin-web/issues/4740

### CODE REVIEW

File: src/elements/emby-textarea/emby-textarea.js [hunk 1]
```diff
@@ -52,16 +52,36 @@ function AutoGrow(textarea, maxLines) {
...
+
+            const newTextValue = textBeforeCursor + '\n' + textAfterCursor;
+            textarea.value = newTextValue;
+
+            const newCursorPosition = currentCursorPosition + 1;
+            textarea.setSelectionRange(newCursorPosition, newCursorPosition);
+
+            autogrowFn();
+        }
+    });
```
Comment (dmitrylyzo) [inline src/elements/emby-textarea/emby-textarea.js:84]: This doubles the new lines and seems unnecessary.
Or maybe I haven't found the steps to reproduce what it is supposed to fix. :man_shrugging:

File: src/elements/emby-textarea/emby-textarea.js [hunk 2]
```diff
@@ -52,16 +52,36 @@ function AutoGrow(textarea, maxLines) {
...
 
     // Call autogrowFn() when textarea's value is changed
-    textarea.addEventListener('input', autogrowFn);
+    textarea.addEventListener('input', function() {
+        autogrowFn();
+        if (textarea.value.trim() === '') {
+            textarea.style.height = 'auto';
+        }
+    });
     textarea.addEventListener('focus', autogrowFn);
```
Comment (dmitrylyzo) [inline src/elements/emby-textarea/emby-textarea.js:68]: Currently, the text area only grows and trimmed only when clearing.
Maybe it is better to trim on `blur`?
```suggestion
    textarea.addEventListener('input', autogrowFn);
    textarea.addEventListener('focus', autogrowFn);
    textarea.addEventListener('blur', () => {
        textarea.style.height = 'auto';
        autogrowFn();
    });
```

`focus` here is redundant, imo. The text area should already be sized correctly. :man_shrugging:

File: src/elements/emby-textarea/emby-textarea.js (hunk 1 above)
Comment (Sukhendu2002) [reply src/elements/emby-textarea/emby-textarea.js:84]: When we add a new line in the text area the text jumps to the top of the screen, it is supposed to fix that.

https://github.com/jellyfin/jellyfin-web/assets/76804228/faa958f1-efe3-41c7-8fb2-9362c0fe19e4

File: src/elements/emby-textarea/emby-textarea.js (hunk 2 above)
Comment (Sukhendu2002) [reply src/elements/emby-textarea/emby-textarea.js:68]: Yea. Got it

File: src/elements/emby-textarea/emby-textarea.js (hunk 1 above)
Comment (dmitrylyzo) [reply src/elements/emby-textarea/emby-textarea.js:83]: It doesn't jump without `keydown` in Chrome and Firefox. :man_shrugging:

Comment (Sukhendu2002) [reply src/elements/emby-textarea/emby-textarea.js:83]: yea

Comment (Sukhendu2002) [reply src/elements/emby-textarea/emby-textarea.js:83]: So, it is pretty much done

Comment (dmitrylyzo) [reply src/elements/emby-textarea/emby-textarea.js:83]: It doubles `\n`.

Comment (Sukhendu2002) [reply src/elements/emby-textarea/emby-textarea.js]: fixed it

Comment (Sukhendu2002) [reply src/elements/emby-textarea/emby-textarea.js]: @dmitrylyzo Can you check it

Comment (Sukhendu2002) [reply src/elements/emby-textarea/emby-textarea.js]: @dmitrylyzo, Can you please review the changes

Comment (dmitrylyzo) [reply src/elements/emby-textarea/emby-textarea.js]: Yes, it does work. I was wondering why this happens and if there is a way without manually manipulating the strings.

Also, I found a trick to fix jumping when reducing the size: https://github.com/angular/material/issues/3070#issuecomment-983418846
We can insert an empty div after the textarea (in `attachedCallback`) and pass it to `AutoGrow`.

Comment (Sukhendu2002) [reply src/elements/emby-textarea/emby-textarea.js]: Hey @dmitrylyzo I tried what you suggested, but it is not working 
This is the code:

```javascript
import './emby-textarea.scss';
import 'webcomponents.js/webcomponents-lite';
import '../emby-input/emby-input';

/**
 * Calculates the vertical padding of the element
 * @param textarea
 * @returns {number}
 */

function calculateOffset(textarea) {
    const style = window.getComputedStyle(textarea, null);
    const props = ['paddingTop', 'paddingBottom'];
    let offset = 0;

    for (const prop of props) {
        offset += parseInt(style[prop], 10);
    }
    return offset;
}

function AutoGrow(textarea, stub, maxLines) {
    const self = this;

    if (maxLines === undefined) {
        maxLines = 999;
    }

    let offset;
    function reset() {
        textarea.rows = 1;
        offset = calculateOffset(textarea);
        self.rows = textarea.rows || 1;
        self.lineHeight = (textarea.scrollHeight / self.rows) - (offset / self.rows);
        self.maxAllowedHeight = (self.lineHeight * maxLines) - offset;
    }

    function autogrowFn() {
        if (!self.lineHeight || self.lineHeight <= 0) {
            reset();
        }
        if (self.lineHeight <= 0) {
            textarea.style.overflowY = 'scroll';
            textarea.style.height = 'auto';
            textarea.rows = 3;
            return;
        }
        let newHeight = 0;

        if ((textarea.scrollHeight - offset) > self.maxAllowedHeight) {
            textarea.style.overflowY = 'scroll';
            newHeight = self.maxAllowedHeight;
        } else {
            textarea.style.overflowY = 'hidden';
            newHeight = textarea.scrollHeight;
        }
        textarea.style.height = newHeight + 'px';
        stub.style.height = newHeight + 'px';
    }

    // Call autogrowFn() when textarea's value is changed
    textarea.addEventListener('input', autogrowFn);
    textarea.addEventListener('focus', autogrowFn);
    textarea.addEventListener('blur', () => {
        textarea.style.height = 'auto';
        autogrowFn();
    });
    textarea.addEventListener('valueset', autogrowFn);
    textarea.addEventListener('keydown', autogrowFn);

    autogrowFn();
}

const EmbyTextAreaPrototype = Object.create(HTMLTextAreaElement.prototype);

let elementId = 0;

if (Object.getOwnPropertyDescriptor && Object.defineProperty) {
    const descriptor = Object.getOwnPropertyDescriptor(HTMLTextAreaElement.prototype, 'value');

    // descriptor returning null in webos
    if (descriptor?.configurable) {
        const baseSetMethod = descriptor.set;
        descriptor.set = function (value) {
            baseSetMethod.call(this, value);

            this.dispatchEvent(new CustomEvent('valueset', {
                bubbles: false,
                cancelable: false
            }));
        };

        Object.defineProperty(HTMLTextAreaElement.prototype, 'value', descriptor);
    }
}

EmbyTextAreaPrototype.createdCallback = function () {
    if (!this.id) {
        this.id = 'embytextarea' + elementId;
        elementId++;
    }
};

EmbyTextAreaPrototype.attachedCallback = function () {
    if (this.classList.contains('emby-textarea')) {
        return;
    }

    this.rows = 1;
    this.classList.add('emby-textarea');

    const parentNode = this.parentNode;
    const label = this.ownerDocument.createElement('label');
    label.innerText = this.getAttribute('label') || '';
    label.classList.add('textareaLabel');

    label.htmlFor = this.id;
    parentNode.insertBefore(label, this);

    // Create a stub element and insert it after the textarea
    const stub = document.createElement('div');
    stub.classList.add('emby-textarea-stub');
    parentNode.insertBefore(stub, this.nextSibling);

    this.addEventListener('focus', function () {
        label.classList.add('textareaLabelFocused');
        label.classList.remove('textareaLabelUnfocused');
    });
    this.addEventListener('blur', function () {
        label.classList.remove('textareaLabelFocused');
        label.classList.add('textareaLabelUnfocused');
    });

    this.label = function (text) {
        label.innerText = text;
    };

    new AutoGrow(this, stub);
};

document.registerElement('emby-textarea', {
    prototype: EmbyTextAreaPrototype,
    extends: 'textarea'
});

```

Comment (Sukhendu2002) [reply src/elements/emby-textarea/emby-textarea.js]: still getting the jumping problem

Comment (dmitrylyzo) [reply src/elements/emby-textarea/emby-textarea.js]: > still getting the jumping problem

On shrinking (deleting text) or on growing (adding text)?
The stub hack is for shrinking. I tried with your `keydown` hack.

Comment (Sukhendu2002) [reply src/elements/emby-textarea/emby-textarea.js]: on growing

Comment (dmitrylyzo) [reply src/elements/emby-textarea/emby-textarea.js]: I have this change with the current state of your branch.
```diff
diff --git a/src/elements/emby-textarea/emby-textarea.js b/src/elements/emby-textarea/emby-textarea.js
index 80e79e115..3189cc22f 100644
--- a/src/elements/emby-textarea/emby-textarea.js
+++ b/src/elements/emby-textarea/emby-textarea.js
@@ -19,7 +19,7 @@ function calculateOffset(textarea) {
     return offset;
 }
 
-function AutoGrow(textarea, maxLines) {
+function AutoGrow(textarea, stub, maxLines) {
     const self = this;
 
     if (maxLines === undefined) {
@@ -52,19 +52,17 @@ function AutoGrow(textarea, maxLines) {
             newHeight = self.maxAllowedHeight;
         } else {
             textarea.style.overflowY = 'hidden';
-            // textarea.style.height = 'auto';
+            stub.style.height = textarea.style.height;
+            textarea.style.height = 'auto';
             newHeight = textarea.scrollHeight/* - offset*/;
         }
         textarea.style.height = newHeight + 'px';
+        stub.style.height = '0px';
     }
 
     // Call autogrowFn() when textarea's value is changed
     textarea.addEventListener('input', autogrowFn);
     textarea.addEventListener('focus', autogrowFn);
-    textarea.addEventListener('blur', () => {
-        textarea.style.height = 'auto';
-        autogrowFn();
-    });
     textarea.addEventListener('valueset', autogrowFn);
     textarea.addEventListener('keydown', function(event) {
         if (event.key === 'Enter') {
@@ -132,6 +130,9 @@ EmbyTextAreaPrototype.attachedCallback = function () {
     label.htmlFor = this.id;
     parentNode.insertBefore(label, this);
 
+    const stub = this.ownerDocument.createElement('div');
+    parentNode.insertBefore(stub, this.nextSibling);
+
     this.addEventListener('focus', function () {
         label.classList.add('textareaLabelFocused');
         label.classList.remove('textareaLabelUnfocused');
@@ -145,7 +146,7 @@ EmbyTextAreaPrototype.attachedCallback = function () {
         label.innerText = text;
     };
 
-    new AutoGrow(this);
+    new AutoGrow(this, stub);
 };
 
 document.registerElement('emby-textarea', {
```

But I noticed that the focus doesn't follow the cursor on `Enter`, when it is at the bottom of the window.

Also, the cursor can hide under the top bar. But this is another story.

**Pull request closed without merging**